src/main/java/org/pac4j/jax/rs/pac4j/JaxRsProfileManager.java

```diff
@@ -0,0 +1,155 @@
+import java.util.LinkedHashMap;
+import java.util.Objects;
+import java.util.Set;
+
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.core.SecurityContext;
+
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.profile.CommonProfile;
+import org.pac4j.core.profile.ProfileHelper;
+import org.pac4j.core.profile.ProfileManager;
+
+/**
+ * 
+ * @author Victor Noel - Linagora
+ * @since 1.0.0
+ *
+ */
+public class JaxRsProfileManager extends ProfileManager<CommonProfile> {
+
+    public JaxRsProfileManager(WebContext context) {
+        super(context);
+
+        ContainerRequestContext requestContext = getJaxRsContext().getRequestContext();
+        SecurityContext original = requestContext.getSecurityContext();
+        if (!(original instanceof Pac4JSecurityContext)) {
+            requestContext.setSecurityContext(new Pac4JSecurityContext(original));
+        }
+    }
+
+    protected JaxRsContext getJaxRsContext() {
+        return (JaxRsContext) this.context;
+    }
+
+    @Override
+    protected LinkedHashMap<String, CommonProfile> retrieveAll(boolean readFromSession) {
+        LinkedHashMap<String, CommonProfile> profiles = super.retrieveAll(readFromSession);
+
+        SecurityContext securityContext = getJaxRsContext().getRequestContext().getSecurityContext();
+        if (securityContext instanceof Pac4JSecurityContext) {
+            ((Pac4JSecurityContext) securityContext).setPrincipal(profiles);
+        }
+
+        return profiles;
+    }
+
+    @Override
+    public void logout() {
+        super.logout();
+
+        SecurityContext securityContext = getJaxRsContext().getRequestContext().getSecurityContext();
+        if (securityContext instanceof Pac4JSecurityContext) {
+            ((Pac4JSecurityContext) securityContext).unsetPrincipal();
+        }
+    }
+
+    private static class Pac4JSecurityContext implements SecurityContext {
+
+        private final SecurityContext original;
+
+        private Principal principal;
+        private Set<String> roles;
+
+        public Pac4JSecurityContext(SecurityContext original) {
+            this.original = original;
+        }
+
+        public void setPrincipal(LinkedHashMap<String, CommonProfile> profiles) {
+            if (profiles != null && !profiles.isEmpty()) {
+                final CommonProfile profile = ProfileHelper.flatIntoOneProfile(profiles).get();
+                principal = new PrincipalImpl(profile.getId());
+                roles = new HashSet<>();
+                profiles.values().stream().forEach(p -> roles.addAll(p.getRoles()));
+            }
+        }
+
+        public void unsetPrincipal() {
+            this.principal = null;
+            this.roles = null;
+        }
+
+        @Override
+        public Principal getUserPrincipal() {
+            if (principal != null) {
+                return principal;
+            } else {
+                return original != null ? original.getUserPrincipal() : null;
```

Comment:
No, the idea is that I replace the `SecurityContext` with our own, but until we are authenticated, I return the original `SecurityContext` values and when we logout, I return also the original `SecurityContext` values.